File: pantry-tracker-backend/database_operations/addListToPantry.py
from extensions import db
from sqlalchemy import text
from flask import Blueprint, jsonify

addListToPantry_bp = Blueprint('addListToPantry', __name__)

@addListToPantry_bp.route('/add-list-to-pantry', methods=['POST'])

def addListToPantry():
    try:
        groceryListIngredients = db.session.execute(text('SELECT aliased_ingredient_name, quantity, unit_id FROM my_grocery_list WHERE quantity > 0')).fetchall()

        if not groceryListIngredients:
            return jsonify({'success': False, 'message': 'No ingredients to add.'}), 200
        
        for ingredient in groceryListIngredients:
            db.session.execute(text('UPDATE my_pantry SET quantity = quantity + :quantity, unit_id = :unit_id WHERE aliased_ingredient_name = :ingredientName'),
            {'quantity': ingredient.quantity, 'unit_id': ingredient.unit_id, 'ingredientName': ingredient.aliased_ingredient_name})
            
        db.session.commit()
        return jsonify({'success': True, 'message': 'Grocery List added to Pantry Successfully!'}), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

File: pantry-tracker-frontend/src/supabase_db_operations/addListToPantry.js
import supabaseClient from "../utils/SupabaseClient";

const addListToPantry = async (userID) => {
    try {
        const { data: namesAndQuantities, error: getNamesAndQuantitiesError } = await supabaseClient
            .from('user_ingredients')
            .select('aliased_ingredient_name, p_quantity, gl_quantity, gl_unit_id')
            .eq('user_id', userID)

        if (getNamesAndQuantitiesError) {
            throw getNamesAndQuantitiesError;
        };
        
        const ingredients = namesAndQuantities.map(ingredient => ({
            name:ingredient.aliased_ingredient_name,
            quantity:ingredient.p_quantity + ingredient.gl_quantity,
            unit_id:ingredient.gl_unit_id
        }));
        
        for (const ingredient of ingredients) {

            const { error: addListToPantryError } = await supabaseClient
                .from('user_ingredients')
                .update({
                    'p_quantity': ingredient.quantity,
                    'p_unit_id': ingredient.unit_id})
                .eq('aliased_ingredient_name', ingredient.name)
                .gt('gl_quantity', 0)
                .eq('user_id', userID);

            if (addListToPantryError) {
                throw addListToPantryError;
            };
        };

        return { success: true };
    }
    catch (error) {
        console.error(error);
        return { success: false };
    };
};

export default addListToPantry;